stdlib/toolchain/Compatibility53/BuiltinProtocolConformances.cpp

```diff
@@ -0,0 +1,136 @@
+  auto func = SWIFT_LAZY_CONSTANT(
+    reinterpret_cast<Fn *>(
+      dlsym(RTLD_DEFAULT, "swift_conformsToProtocol")));
+  return func(type, protocol);
+}
+
+#define TUPLE_EQUATABLE_WT SYMBOL("_swift_tupleEquatable_wt")
+
+// Define the conformance descriptor for tuple Equatable. We do this in
+// assembly to work around relative reference issues.
+__asm(
+  "  .section __DATA,__data\n"
+  "  .globl " TUPLE_EQUATABLE_CONF "\n"
+  "  .p2align 2\n"
+  TUPLE_EQUATABLE_CONF ":\n"
+  // This is an indirectable relative reference to the Equatable protocol
+  // descriptor. However, this is 0 here because the compatibility libraries
+  // can't have a dependency on libswiftCore (which is where Equatable live).
+  "  .long 0\n"
+  // 769 is the MetadataKind::Tuple
+  "  .long 769\n"
+  // This is a direct relative reference to the witness table defined below.
+  "  .long ((" TUPLE_EQUATABLE_WT ") - (" TUPLE_EQUATABLE_CONF ")) - 8\n"
+  // 32 are the ConformanceFlags with the type reference bit set to MetadataKind.
+  "  .long 32\n"
+);
+
+extern const ProtocolConformanceDescriptor _swift_tupleEquatable_conf;
+
+// Due to the fact that the compatibility libraries can't have a hard
+// dependency to libswiftCore (which is where the Equatable protocol desciptor
+// live), we have to manually implant this before calling any user code.
+__attribute__((constructor))
+void _implaceEquatableDescriptor() {
+  auto tupleEquatableConf = const_cast<int32_t *>(
+    reinterpret_cast<const int32_t *>(&_swift_tupleEquatable_conf));
+  auto equatable = reinterpret_cast<const int32_t *>(getEquatableDescriptor());
+
+  // This is an indirectable pointer.
+  *tupleEquatableConf = intptr_t(equatable) - intptr_t(tupleEquatableConf);
+}
+
+template<unsigned int NumWitnesses>
+struct _WitnessTable {
+  const ProtocolConformanceDescriptor *Conformance;
+  const void *Witnesses[NumWitnesses];
+};
+
+SWIFT_RUNTIME_EXPORT
+const _WitnessTable<1> _swift_tupleEquatable_wt = {
+  &_swift_tupleEquatable_conf,
+  {
+    reinterpret_cast<void *>(_swift_tupleEquatable_equals)
+  }
+};
+
+SWIFT_RUNTIME_EXPORT SWIFT_CC(swift)
+bool swift::_swift_tupleEquatable_equals(OpaqueValue *tuple1,
+                                         OpaqueValue *tuple2,
+                                         SWIFT_CONTEXT Metadata *swiftSelf,
+                                         Metadata *Self, void *witnessTable) {
+  auto tuple = cast<TupleTypeMetadata>(Self);
+
+  // Loop through all elements, and check if both tuples element is equal.
+  for (size_t i = 0; i != tuple->NumElements; i += 1) {
+    auto elt = tuple->getElement(i);
+
+    // Ensure we actually have a conformance to Equatable for this element type.
+    auto equatable = getEquatableDescriptor();
+    auto conformance = conformsToProtocol(elt.Type, equatable);
+
+    // If we don't have a conformance then something somewhere messed up in
+    // deciding that this tuple type was Equatable...??
+    if (!conformance)
+      abort();
```

Comment:
Is there a version of something like `_internalInvariant()` for compatibility libs where you can at least include a message? Like `llvm_unreachable()` (ref: https://llvm.org/docs/CodingStandards.html#assert-liberally) if the Swift compiler headers don't have something more Swift-runtime-specific?

Reply:
I think you can call `crash` or `fatalError`, which is defined [here](https://github.com/apple/swift/blob/ad0a8d31217455bb8878708897a44b1496290cdb/include/swift/Runtime/Debug.h).